sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableInserter.java

```diff
@@ -273,6 +296,137 @@ public void insertAll(TableReference ref, List<TableRow> rowList,
+    job.setJobReference(jobRef);
+    JobConfiguration config = new JobConfiguration();
+    config.setLoad(loadConfig);
+    job.setConfiguration(config);
+
+    Exception lastException = null;
+    Sleeper sleeper = Sleeper.DEFAULT;
+    BackOff backoff = new AttemptBoundedExponentialBackOff(
+        MAX_LOAD_JOB_RPC_ATTEMPTS, INITIAL_LOAD_JOB_RPC_BACKOFF_MILLIS);
+    do {
+      try {
+        client.jobs().insert(projectId, job).execute();
+        return; // SUCCEEDED
+      } catch (GoogleJsonResponseException e) {
+        if (errorExtractor.itemAlreadyExists(e)) {
+          return; // SUCCEEDED
+        }
+        // ignore and retry
+        LOG.warn("Ignore the error and retry inserting the job.", e);
+        lastException = e;
+      } catch (IOException e) {
+        // ignore and retry
+        LOG.warn("Ignore the error and retry inserting the job.", e);
+        lastException = e;
+      }
+    } while (nextBackOff(sleeper, backoff));
+    throw new IOException(
+        String.format(
+            "Unable to insert job: {}, aborting after {} retries.",
+            jobId, MAX_LOAD_JOB_RPC_ATTEMPTS),
+        lastException);
+  }
+
+  @VisibleForTesting
+  enum Status {
+    SUCCEEDED,
+    FAILED,
+    UNKNOWN,
+  }
+
+  private Status pollJobStatus(String projectId, String jobId) throws InterruptedException {
+    BackOff backoff = new AttemptBoundedExponentialBackOff(
+        MAX_LOAD_JOB_POLL_RETRIES, INITIAL_LOAD_JOB_POLL_BACKOFF_MILLIS);
+    return pollJobStatus(projectId, jobId, Sleeper.DEFAULT, backoff);
+  }
+
+  @VisibleForTesting
+  Status pollJobStatus(
+      String projectId,
+      String jobId,
+      Sleeper sleeper,
+      BackOff backoff) throws InterruptedException {
+    do {
+      try {
+        JobStatus status = client.jobs().get(projectId, jobId).execute().getStatus();
+        if (status != null && status.getState() != null && status.getState().equals("DONE")) {
+          if (status.getErrorResult() != null) {
+            return Status.FAILED;
+          } else if (status.getErrors() != null && !status.getErrors().isEmpty()) {
+            return Status.FAILED;
+          } else {
+            return Status.SUCCEEDED;
+          }
+        }
+        // The job is not DONE, wait longer and retry.
+      } catch (IOException e) {
+        // ignore and retry
+        LOG.warn("Ignore the error and retry polling job status.", e);
+      }
+    } while (nextBackOff(sleeper, backoff));
+    LOG.warn("Unable to poll job status: {}, aborting after {} retries.",
+        jobId, MAX_LOAD_JOB_POLL_RETRIES);
+    return Status.UNKNOWN;
+  }
+
+  /**
+   * Identical to {@link BackOffUtils#next} but without checked IOException.
+   * @throws InterruptedException
+   */
+  private boolean nextBackOff(Sleeper sleeper, BackOff backoff) throws InterruptedException {
+    try {
+      return BackOffUtils.next(sleeper, backoff);
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
```

Comment:
`throw new RuntimeException("Unexpected error encountered during sleep", e);`

I have no idea why that function throws. None of its uses that I could find do so.